provider/kubernetes/kubernetes.go

```diff
@@ -677,3 +706,64 @@ func getRateLimit(i *extensionsv1beta1.Ingress) *types.RateLimit {
 
 	return rateLimit
 }
+
+func getServicesPercentageWeights(i *extensionsv1beta1.Ingress) (map[string]*percentageValue, error) {
+	rawPercentageWeights := getStringValue(i.Annotations, annotationKubernetesPercentageWeights, "")
+
+	percentageWeight := make(map[string]string)
+
+	if err := yaml.Unmarshal([]byte(rawPercentageWeights), percentageWeight); err != nil {
+		return nil, err
+	}
+
+	servicesPercentageWeights := make(map[string]*percentageValue)
+	for serviceName, percentageStr := range percentageWeight {
+		percentageValue, err := newPercentageValueFromString(percentageStr)
+		if err != nil {
+			return nil, fmt.Errorf("invalid percentage value %q in ingress %s: %v", percentageStr, i.Name, err)
+		}
+
+		servicesPercentageWeights[serviceName] = percentageValue
+	}
+	return servicesPercentageWeights, nil
+}
+
+func getLeftFraction(k8sClient Client, namespace string, paths []extensionsv1beta1.HTTPIngressPath, servicesPercentageWeights map[string]*percentageValue) (map[string]*percentageValue, map[string]int, error) {
+	leftFractionPercentages := make(map[string]*percentageValue)
+	leftFractionInstanceCounts := make(map[string]int)
+	oneHundredPercentageValue := newPercentageValueFromFloat64(1)
+
+	for _, pa := range paths {
+		leftFractionPercentages[pa.Path] = oneHundredPercentageValue
+		leftFractionInstanceCounts[pa.Path] = 0
+	}
+
+	for _, pa := range paths {
+		endpoints, exist, err := k8sClient.GetEndpoints(namespace, pa.Backend.ServiceName)
```

Comment:
Yes, I agree with you. This function is not really to simplify the abstraction but to make the loop logic over ingresses looks short as far as I think.
 
> Is there a chance we can inject the number of subset addresses

What do you mean by "inject"? To retrieve the endpoint at the very begin of iterating paths of ingress? Since that we are using informers to do the actual "get" or "list". The impact of performance should be neglectable.